adverbio.1
.\" (c) 2014 Richard Kettlewell
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.TH ADVERBIO 1 "June 26, 2001"
.SH NAME
adverbio \- I/O redirection with adverbial commands
.SH SYNOPSIS
.B adverbio
.RI [ options ]
.RB [ -- ]
.IR command ...
.SH DESCRIPTION
.B adverbio
performs one or more I/O redirection operations and then executes the
command via
.BR execvp (3).
It is intended to be used where shell I/O redirection is
inappropriate, either because the shell is not in use, or because it
would perform redirection operations at the wrong moment.
.PP
For example, consider the command "really", which executes its
arguments as root.  If you want to redirect the output of such a
command then the naive approach:
.PP
.nf
really ls > /root/file
.fi
.PP
does not work - the mortal user's shell tries to perform the
redirection, and this fails.  It is necessary to reinvoke the shell to
get it to perform the redirection at the right point:
.PP
.nf
really sh \-c 'ls > /root/file'
.fi
.PP
However this is unacceptable for general use, as it requires shell
commands to be shell-quoted - and so any argument strings must be
quoted twice!  This is messy and error prone.
.PP
With
.B adverbio
the above becomes:
.PP
.nf
really adverbio \-o /root/file \-\- ls
.fi
.PP
which is slightly verbose, but requires no quoting.
.SH OPTIONS
\fBadverbio\fP accepts the following options:
.TP
\fB-o\fR \fIpath\fI, \fB--stdout\fR \fIpath\fI
Redirects standard output to output file PATH.
.TP
\fB-e\fR \fIpath\fI, \fB--stderr\fR \fIpath\fI
Redirects standard error to output file PATH.
.TP
\fB-i\fR \fIpath\fI, \fB--stdin\fR \fIpath\fI
Redirects standard input from input file PATH.
.TP
\fB-O\fR \fIfd\fB:\fIpath\fI, \fB--output\fR \fIfd\fB:\fIpath\fI
Redirects file descriptor FD to output file PATH, destroying any
previous meaning for FD.
.TP
\fB-I\fR \fIfd\fB:\fIpath\fI, \fB--input\fR \fIfd\fB:\fIpath\fI
Redirects file descriptor FD from input file PATH, destroying any
previous meaning for FD.
.TP
\fB-r\fR \fIfd\fB:\fIofd\fI, \fB--redirect\fR \fIfd\fB:\fIofd\fI
Makes file descriptor FD a copy of existing file descriptor OFD,
destroying any previous meaning for FD.
.TP
\fB-C\fR \fIfd\fB, \fB--close\fR \fIfd\fB
Closes file descriptor FD, which must be open.
.TP
\fB-a\fR, \fB--append\fR
Specifies that all output files opened after this option will be
opened in "append" mode.  If the file does not exist it will be
created.  All writes will be appended to the file.
.TP
\fB-c\fR, \fB--clobber\fR
Specifies that all output files opened after this option will be
opened in "clobber" mode.  If the file exists it will be truncated.
.TP
\fB-s\fR, \fB--safe\fR
Specifies that all output files opened after this option will be
opened in "safe" mode.  If the file exists then an error will be
reported.  This is the default.
.TP
\fB-S\fR, \fB--sync\fR
Specifies that all output files opened after this option will be
opened for synchronous I/O.
.TP
\fB-h\fR, \fB--help\fR
Show summary of options.
.TP
\fB-V\fR, \fB--version\fR
Show version of program.
.SH AUTHOR
Richard Kettlewell <rjk@greenend.org.uk>.


ChangeLog
2005-02-07 17:17:21 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * adverbio.c: -O FD:PATH where the open(PATH, ...) returned a value
          matching FD would unhelpfuly close FD.  Problem reported by Colin
          Watson.
2005-02-07 17:14:08 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * with-lock.c: bogus string literal
2003-07-19 17:35:12  Richard Kettlewell  <rjk@greenend.org.uk>

        * check-tested-everything, test-*: check that every executable has
          been tested
        
        * configure.ac: quieten autoconf
        
        * debug.c, utils.h, run-repeatedly.c: new debugging interface
        
        * test-run-repeatedly: fix test
        
2003-07-19 11:26:29  Richard Kettlewell  <rjk@greenend.org.uk>

        * prepare: don't configure if asked not to
        
        * test-run-repeatedly: tests
        
2003-03-08  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.1: added some examples
	
2002-11-28  Richard Kettlewell  <rjk@greenend.org.uk>

	* iobuffer.c: use a single big circular buffer rather than lots of
	small buffers.  This makes more predictable and efficient use of
	memory than the previous strategy.

	Also change --max-buffer/-m to --buffer/-b since it is now the
	exact buffer size.

2002-11-23  Richard Kettlewell  <rjk@greenend.org.uk>

	* More test infrastructure.  Various tests improved.

	* run-repeatedly.c: underscore in filenames is allowed (to match
	run-parts)

2002-11-22  Richard Kettlewell  <rjk@greenend.org.uk>

	* various changes to build on FreeBSD.  daemon (at least) doesn't
	seem to work properly there though.

2002-11-21  Richard Kettlewell  <rjk@greenend.org.uk>

	* alarm.c: report failed setpgid correctly

	* lots of <syscall>_e functions to keep error handling out of the
	main code.  Convert a couple of malloc/realloc calls to the x-
	versions too (for the same reason).

	* more tests

2002-11-20  Richard Kettlewell  <rjk@greenend.org.uk>

	* test-daemon: more tests, including for some of the new features

	* logfds.c: adjust option letters so daemon & logfds are
	consistent.

	* daemon.c: saner error output handling - essentially build part
	of logfds into daemon.

2002-11-18  Richard Kettlewell  <rjk@greenend.org.uk>

	* anagrams.c: use dynamic memory allocation rather than fixed-size
	buffers.  Now there should be no size limits other than available
	memory.

	Also, now you can specify multiple word lists, and duplicate words
	don't cause duplicate anagrams.

2002-11-17  Richard Kettlewell  <rjk@greenend.org.uk>

	* new programs: iobuffer, anagrams

2002-11-11  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.c: fix message reported when a subprocess gets a signal

	* alarm: new --verbose option

	* new programs: logfds, bind/accept/connect-socket, pidfile,
	run-as, daemon, run-repeatedly, with-lock

	* switch to new autoconf &c

2001-09-09  Richard Kettlewell  <rjk@greenend.org.uk>

	* configure.in: changed version to 0.6.0

	* added "daemon" program.

2001-08-12  Richard Kettlewell  <rjk@greenend.org.uk>

	* merged "inplace" and "adverbio" into a single package



adverbio.c
/* 
   adverbio - I/O redirection with adverbial commands

   Copyright (C) 2001 Richard Kettlewell

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <config.h>
#include <getopt.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>

#include "utils.h"

/* Option flags and variables */

static struct option const long_options[] =
{
  { "help", no_argument, 0, 'h' },
  { "version", no_argument, 0, 'V' },
  { "stdout", required_argument, 0, 'o' },
  { "stdin", required_argument, 0, 'i' },
  { "stderr", required_argument, 0, 'e' },
  { "output", required_argument, 0, 'O' },
  { "input", required_argument, 0, 'I' },
  { "redirect", required_argument, 0, 'r' },
  { "close", required_argument, 0, 'C' },
  { "append", no_argument, 0, 'a' },
  { "clobber", no_argument, 0, 'c' },
  { "safe", no_argument, 0, 's' },
  { "sync", no_argument, 0, 'S' },
  { 0, 0, 0, 0}
};

/* write a usage message to FP and exit with the specified status */

static void __attribute__((noreturn)) usage(FILE *fp, int exit_status) {
  if(fprintf(fp, "Usage:\n"
	     "\n"
	     "  adverbio [options] -- command...\n"
	     "\n"
	     "Options:\n"
	     "\n"
	     "  -o FILENAME, --stdout FILENAME        Redirect stdout\n"
	     "  -e FILENAME, --stderr FILENAME        Redirect stderr\n"
	     "  -i FILENAME, --stdin FILENAME         Redirect stdin\n"
	     "  -O FD:FILENAME, --output FD:FILENAME  Redirect FD to a file\n"
	     "  -I FD:FILENAME, --input FD:FILENAME   Redirect FD from a file\n"
	     "  -r FD1:FD2, --redirect FD1:FD2        Redirect FD1 to FD2\n"
	     "  -C FD, --close FD                     Close FD\n"
	     "  -a, --append                          Output in append mode\n"
	     "  -c, --clobber                         Output in clobber mode\n"
	     "  -s, --safe                            Output in safe mode\n"
#if O_SYNC
	     "  -S, --sync                            Output synchronously\n"
#endif
	     "  -h, --help                            Usage message\n"
	     "  -V, --version                         Version number\n"
	     "\n") < 0)
    fatale("output error");
  exit(exit_status);
}

/* redirect TARGET_FD to file PATH, using open() flags FLAGS */

static void redirect_to_file(int target_fd, const char *path, int flags) {
  int fd;

  fd = open_e(path, flags, 0666);
  if(fd != target_fd) {
    dup2_e(fd, target_fd);
    close_e(fd);
  }
}

/* parse a FD:FILENAME argument */

static void parse_fd_filename(const char *arg,
			      int *fdp,
			      const char **filenamep) {
  int n = strspn(arg, "0123456789");

  if(n == 0 || arg[n] != ':')
    usage(stderr, 1);
  *fdp = atoi(arg);
  *filenamep = arg + n + 1;
}

/* parse a FD:FD argument */

static void parse_fd_fd(const char *arg,
			int *fd1p,
			int *fd2p) {
  int n = strspn(arg, "0123456789");

  if(n == 0 || arg[n] != ':'
     || arg[n + 1 + strspn(arg + n + 1, "0123456789")] != 0)
    usage(stderr, 1);
  *fd1p = atoi(arg);
  *fd2p = atoi(arg + n + 1);
}


int main(int argc, char **argv) {
  int n;
  int oflags = O_CREAT|O_EXCL;
  int syncflag = 0;
  int fd1, fd2;
  const char *filename;

  setprogname(argv[0]);

  while((n = getopt_long(argc, argv, 
			 "hVo:i:e:I:O:r:C:acsS",
			 long_options, (int *) 0)) >= 0) {
    switch(n) {
    case 'V':
      printf("adverbio %s\n", VERSION);
      return 0;

    case 'h':
      usage(stdout, 0);
      return 0;

    case 'o':
      redirect_to_file(1, optarg, O_WRONLY | oflags | syncflag);
      break;

    case 'e':
      redirect_to_file(2, optarg, O_WRONLY | oflags | syncflag);
      break;

    case 'i':
      redirect_to_file(0, optarg, O_RDONLY);
      break;

    case 'I':
      parse_fd_filename(optarg, &fd1, &filename);
      redirect_to_file(fd1, filename, O_RDONLY);
      break;
      
    case 'O':
      parse_fd_filename(optarg, &fd1, &filename);
      redirect_to_file(fd1, filename, O_WRONLY | oflags | syncflag);
      break;
      
    case 'r':
      parse_fd_fd(optarg, &fd1, &fd2);
      dup2_e(fd2, fd1);
      break;

    case 'C':
      if(!*optarg
	 || optarg[strspn(optarg, "0123456789")] != 0)
	usage(stderr, 1);
      close_e(atoi(optarg));
      break;
      
    case 'c':
      oflags = O_CREAT|O_TRUNC;
      break;

    case 'a':
      oflags = O_CREAT|O_APPEND;
      break;

    case 's':
      oflags = O_CREAT|O_EXCL;
      break;

    case 'S':
#ifdef O_SYNC
      syncflag = O_SYNC;
#else
      error("warning: -S not supported on this platform");
#endif
      break;
      
    default:
      usage(stderr, 1);
    }
  }

  if(optind >= argc)
    usage(stderr, 1);

  execvp(argv[optind], argv + optind);

  fatale("error executing %s", argv[optind]);
}

/*
Local Variables:
c-basic-offset:2
comment-column:40
End:
*/
